packages/server/utils/useOAuthScopeValidation.ts

```diff
@@ -0,0 +1,59 @@
+import {GraphQLError, getOperationAST} from 'graphql'
+import type {Plugin} from 'graphql-yoga'
+import type {ServerContext, UserContext} from '../yoga'
+
+/**
+ * GraphQL Yoga plugin that validates tokens have the correct scope
+ * when executing GraphQL operations via the OAuth 2.0 API
+ *
+ * - Queries require 'graphql:query' scope
+ * - Mutations require 'graphql:mutation' scope
+ * - Super users bypass this validation
+ */
+export const useOAuthScopeValidation = (): Plugin<ServerContext & UserContext> => {
+  return {
+    onExecute({args}) {
+      return {
+        onExecuteDone() {
```

Comment:
-2 check for permission before doing the mutation, not after

Reply:
i think what you're looking for is `permissions.ts`. add per-operation validation?